# NumPy and Pandas

While base Python has many useful tools for many purposes, open-source external libraries greatly expand Python's uses. Here, we will discuss the two of the most popular libraries: **NumPy** and **Pandas**. Both of these are great for data analysis.

## NumPy

NumPy (Numerical Python) is a critical library for manipulating numbers, performing matrix operations, and mathematics in general. 

To use this library, we first have to **import** it with the keyword `import`.

In [21]:
import numpy

We now have access to the various tools and functions that NumPy has to offer. The foundation of NumPy is the array, a data structure for grouping numbers designed for math. 

The simplest array is a single dimensional vector, essentially a Python list that we can do easy math with. To make an array, we tend to create a list and convert it to an array with `numpy.array()`.

In [22]:
my_list = [1, 2, 3]

my_array = numpy.array(my_list)
print(my_array)

[1 2 3]


You can always do simple math operations between a number (int or float) and an array.

In [41]:
20 * my_array

array([20, 40, 60])

If we have arrays of the same length, we can do the same operations on them between elements in the same positions.

In [24]:
your_array = numpy.array( [3, 2, 1] ) # you can create the list within the array() function

my_array + your_array

array([4, 4, 4])

In [25]:
my_array * your_array

array([3, 4, 3])

#### Bonus
It is common practice in Python to use `import numpy as np` when importing NumPy. This allows you to only need to type `np.` when using a tool within NumPy, which is a bit less clunky and faster.

In [26]:
import numpy as np

np.array([]) # a blank numpy array

array([], dtype=float64)

You could technically import NumPy as any variable name, but please avoid this to avoid confusion.

In [27]:
### DONT RUN THIS
# import numpy as pandas

## 2D Array

NumPy arrays really come into their own when they're used as matrices. Let's first make a 3 x 3 array. To do this, we will call `numpy.array()` with a list that contains other lists, also called a **nested list**. 

In [30]:
numpy.array( [[1, 2, 3],
              [4, 5, 6],
              [7, 8, 9]] )

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

We can get the dimensions of the array by checking the `.shape` attribute.

In [42]:
a = numpy.array( [[1, 1, 1],
                  [1, 1, 1],
                  [1, 1, 1]] )

a.shape

(3, 3)

Just like the single dimensional array, you can use the standard math operators between 2D arrays, though they have to be of the same shape.

In [43]:
b = numpy.array( [[10, 10, 10],
                  [10, 10, 10],
                  [10, 10, 10]] )
a + b 

array([[11, 11, 11],
       [11, 11, 11],
       [11, 11, 11]])

Like a 1D vector, you can also do math operations with a single number.

In [44]:
b - 20

array([[-10, -10, -10],
       [-10, -10, -10],
       [-10, -10, -10]])

NumPy comes with many tools to do various more complicated math operations as well. For instance, `numpy.matmul` can be used for matrix multiplication. 

In [47]:
numpy.matmul(a, b)

array([[30, 30, 30],
       [30, 30, 30],
       [30, 30, 30]])

Here is non-exhaustive list of other useful operations you can calculate with NumPy. Many of them use the submodule `linalg` that specializes in linear algebra operations.
- Mean: `numpy.mean()`
- Median: `numpy.median()`
- Maximum: `numpy.max()`
- Minimum: `numpy.min()`
- Standard deviation: `numpy.std()`
- Variance: `numpy.var()`
- Dot product: `numpy.dot()`
- Determinant: `numpy.linalg.det()`
- Vector/matrix norm: `numpy.linalg.norm()`
- Matrix rank: `numpy.linalg.det()`
- Matrix inverse: `numpy.linalg.inv()`
- Eigenvalues/eigenvectors: `numpy.linalg.eig()`
- Solutions to linear equations: `numpy.linalg.solve()`

For full usage of these functions and more, please visit the [NumPy reference manual](https://numpy.org/doc/stable/reference/routines.linalg.html).

### Indexing and slicing in NumPy

Selecting a value in a 1D array is just like indexing in a Python list. If the array has a length of 4, indexes begin at 0 and end at 3. 

In [49]:
x = np.array([3.2, 4.1, 5.6, 8.3])
x[3]

8.3

2D arrays can be indexed in a similar manner with separate column index and row index -> array[row, col]. Both column and row numbers begin with 0.

![array indexing](https://swcarpentry.github.io/python-novice-inflammation/fig/python-zero-index.svg)
*Credit to [Software Carpentry](https://swcarpentry.github.io/python-novice-inflammation/02-numpy/index.html)*.

In [54]:
y = numpy.array([[1., 2., 3., 4.], # adding the decimal makes them floats
              [5., 6., 7., 8.]]) 
y[1,3] # returns the last value (8.0)

8.0

Also like lists, we can use **negative indexing** to get the last values of a column and/or row.

In [57]:
y[-1,-1] # last row, last column

8.0

We can also using **slicing** to return portions of an array -> `array[i:j]`. Slicing is **inclusive** for the first index (`i`) and **exclusive** for the last index (`j`). `array[i:j]` returns values from `i` to `j-1`. 

In [56]:
print(x[0:2]) # returns values from 0 to 1

array([3.2, 4.1])

We can use this for 2D arrays, as well. We can slice rows, columns, or both at once.

In [64]:
z = numpy.array([[1, 2, 3, 4],
                 [5, 6, 7, 8],
                 [9,10,11,12],
                 [13,14,15,16]])

print(z[0:2, 2]) # first 2 rows of the third column
print(z[1:3, 0:3]) # rows 2 and 3 for columns 1-3

[3 7]
[[ 5  6  7]
 [ 9 10 11]]


#### Question: Slicing

What happens when you slice but do not include the first index (`i`), the last index (`j`), or include neither?

In [ ]:
### try it out:

### Other uses of NumPy

#### Random numbers

If you are working with probability distributions, stochastic processes, or have any use for random number generators, `numpy.random` is a useful module of NumPy. 

## Pandas

In [29]:
import pandas as pd

### Read data
With `pandas` imported, we can read in .csv files with the `pandas` function `read_csv()`.

In that function, we can specify the file we want to use with a URL or with the path to a local file as a string.

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/DeisData/python/master/gapminder.csv") # read in data

Our data is now saved as a data frame in Python as the variable `df`. With the data now in the environment, we can take a look at the first few rows with `df.head()`.

In [3]:
df.head()

country  year region  population  life_expectancy  age5_surviving  \
0  Afghanistan  1800   Asia   3280000.0            28.21          53.142   
1  Afghanistan  1810   Asia   3280000.0            28.11          53.002   
2  Afghanistan  1820   Asia   3323519.0            28.01          52.862   
3  Afghanistan  1830   Asia   3448982.0            27.90          52.719   
4  Afghanistan  1840   Asia   3625022.0            27.80          52.576   

   babies_per_woman  gdp_per_capita  gdp_per_day  
0               7.0           603.0     1.650924  
1               7.0           604.0     1.653662  
2               7.0           604.0     1.653662  
3               7.0           625.0     1.711157  
4               7.0           647.0     1.771389

We can see that this data frame has several different columns, with information about countries and demography

## Summarize data frame

It is important to understand the data we are working with before we begin analysis. First, let's look at the dimenions of the data frame using `df.shape`. It gives the number of rows by the number of columns.

In [4]:
df.shape

(14740, 9)

This shows that our data frame has 14740 rows by 9 columns.

We can also use `df.columns` to display the column names.

In [5]:
df.columns

Index(['country', 'year', 'region', 'population', 'life_expectancy',
       'age5_surviving', 'babies_per_woman', 'gdp_per_capita', 'gdp_per_day'],
      dtype='object')

### Categorical variables
Next, let's summarize the categorical, non-numerical variables. For instance, we can identify how many unique regions we have in the data set.

First, to select a column, we use the notation `df['COLUMN_NAME']`.

In [6]:
df['region']

0          Asia
1          Asia
2          Asia
3          Asia
4          Asia
          ...  
14735    Africa
14736    Africa
14737    Africa
14738    Africa
14739    Africa
Name: region, Length: 14740, dtype: object

To identify unique entries in this column, we can use the `pd.unique()` function. 

In [7]:
pd.unique( df['region'] )

array(['Asia', 'Europe', 'Africa', 'America'], dtype=object)

The `countries` column has many unique values, so we'll just use the `len()` function to see how many unique countries we have.

In [8]:
len( pd.unique( df['country'] ) ) # this is called nesting functions -> calling functions within other functions

182

### Numerical variables

Numerical columns can be summarized in several ways. Let's find the mean first.

To make things simpler, we'll just do calculations on the `population`, `life_expectancy`, and `babies_per_woman` columns. We can put those names in an arrangement called a`list` and then specify that list for the columns.

In [9]:
num_cols = [ 'population', 'life_expectancy', 'babies_per_woman' ] # numerical columns

df[num_cols]

population  life_expectancy  babies_per_woman
0       3280000.0            28.21              7.00
1       3280000.0            28.11              7.00
2       3323519.0            28.01              7.00
3       3448982.0            27.90              7.00
4       3625022.0            27.80              7.00
...           ...              ...               ...
14735  14255592.0            51.60              3.64
14736  14565482.0            54.20              3.56
14737  14898092.0            55.70              3.49
14738  15245855.0            57.00              3.41
14739  15602751.0            59.30              3.35

[14740 rows x 3 columns]

With this set of columns, we can run `.mean()` to find the mean of each column.

In [10]:
df[num_cols].mean() # returns the mean of each column

population          2.252933e+07
life_expectancy     5.683453e+01
babies_per_woman    4.643472e+00
dtype: float64

If we want a larger variety of summary statistics, we can use the `.describe()` method.

In [11]:
df[num_cols].describe()

population  life_expectancy  babies_per_woman
count  1.474000e+04     14740.000000      14740.000000
mean   2.252933e+07        56.834526          4.643472
std    9.307143e+07        15.868464          1.994833
min    2.128000e+03         4.000000          1.130000
25%    8.990308e+05        44.230000          2.630000
50%    4.063978e+06        60.080000          5.060000
75%    1.218722e+07        70.380000          6.440000
max    1.376049e+09        83.300000          9.220000

We can also break down subgroupings of our data with the method `.groupby()`.

In [12]:
grouped_data = df.groupby('region')
grouped_data['population'].describe()

count          mean           std      min         25%        50%  \
region                                                                        
Africa   4293.0  9.181313e+06  1.655128e+07  12522.0   996331.00  3457113.0   
America  2673.0  1.667833e+07  4.411806e+07  24000.0   331799.00  2843246.0   
Asia     4212.0  4.604245e+07  1.658010e+08   2128.0   512028.25  4011309.5   
Europe   3562.0  1.520351e+07  2.463153e+07  61428.0  2308682.00  5186801.5   

                 75%           max  
region                              
Africa    9901052.00  1.822020e+08  
America  10061519.00  3.217736e+08  
Asia     19517390.25  1.376049e+09  
Europe   10638884.75  1.484358e+08

### Accessing rows and specific entries

You can also to access a specific row using `df.loc[ROW, :]`. The colon specifies to select all columns for that row number.

In [13]:
df.loc[0, :] # the first row

country             Afghanistan
year                       1800
region                     Asia
population            3280000.0
life_expectancy           28.21
age5_surviving           53.142
babies_per_woman            7.0
gdp_per_capita            603.0
gdp_per_day            1.650924
Name: 0, dtype: object

We can use `.loc` to find the value of specific entries, as well.

In [14]:
df.loc[0, 'country']

'Afghanistan'

#### Question
Print out the summary statistics for columns `age5_surviving`, `gdp_per_day`, and `gdp_per_capita`.

In [15]:
### your code below:


## Manipulate data 

### Subset by row

Sometimes, we want to create a subset of the main data frame based on certain conditions. We do this by using `df.loc` and specifying a condition for the rows. 

Below, we take all of the rows where `babies_per_woman` is greater or equal to 4 with `df['babies_per_woman'] >= 4` and assign this to a new data frame.

To check that this was done correctly, we can look at the minimum of the `babies_per_woman` column in the new data frame with  `.min()`.

In [16]:
# take all rows where babies_per_woman is greater or equal to 4 and make a new data frame
df_4 = df.loc[df['babies_per_woman'] >= 4, :]
df_4['babies_per_woman'].min()

4.0

We can also subset with categorical variables. Here, we take all rows where the country is Hungary. 

In [17]:
df_hungary = df.loc[df['country'] == 'Hungary', :]
pd.unique(df_hungary['country'])

array(['Hungary'], dtype=object)

### Math

If we multiply a data frame by a single number, each value in the column will be muliplied by that value.

In [18]:
df['babies_per_woman'] * 1000

0        7000.0
1        7000.0
2        7000.0
3        7000.0
4        7000.0
          ...  
14735    3640.0
14736    3560.0
14737    3490.0
14738    3410.0
14739    3350.0
Name: babies_per_woman, Length: 14740, dtype: float64

We can also do math between columns, since they have the same length. Elements of the same row are added, substacted, multiplied, or divided. 

Here, we subtract the `life_expectancy` column from the `age5_surviving` column and assign it to a new column called `life_difference`. 

In [19]:
df['life_difference'] = df['age5_surviving'] - df['life_expectancy'] 
print(df['life_difference'])

0        24.932
1        24.892
2        24.852
3        24.819
4        24.776
          ...  
14735    39.200
14736    37.130
14737    35.970
14738    34.900
14739    32.740
Name: life_difference, Length: 14740, dtype: float64


This new column is now reflected in the data frame. 

In [20]:
print(df.columns)

Index(['country', 'year', 'region', 'population', 'life_expectancy',
       'age5_surviving', 'babies_per_woman', 'gdp_per_capita', 'gdp_per_day',
       'life_difference'],
      dtype='object')


#### Question 

Create a subset of data from Lithuania. 

Within that subset, calculate the mean GDP per 1000 people across entries.

*Hint: Multiply per capita GDP by 1000.*

In [ ]:
### Your code here:


## Resources

This lesson is adapted from 
[Software Carpentry](http://swcarpentry.github.io/python-novice-gapminder/design/).